[Misc] Add a script to benchmark compilation time

desertfire opened the pull request:
Summary: Adds a benchmarking script for measuring torch.compile's cold and warm compilation times. For cold compilation, the script redirects both vLLM and torch.compile caching to a temporary directory on each iteration. For warm compilation, it uses the default caching directories.

The script sets VLLM_ENABLE_V1_MULTIPROCESSING=0 and runs each iteration in a separate subprocess to track the compilation time.

Usage: `vllm bench compilation --model YOUR_MODEL`

ProExpertProg reviewed (commented):
Thanks for doing this! A couple of thoughts:
- Let's rename the script from `bench compilation` to `bench startup`: we're measuring (and reporting) the total startup anyway. We also cache model info and we might cache other things in the future, so cold/warm distinction will apply to more than compilation as well.
- It would be nice if we can at least add some form of "total startup" reporting for all benchmarking scripts (can be warm-only). That's because devs often only report results from a single benchmark and I want to start tracking startup in PRs, but devs are less likely to run a separate script just for startup.
- Could you also add a `vllm bench sweep startup` command? I want to move more people over to the new `vllm bench sweep serve` command which sweeps over combinations of serving parameters and request rates, and reusing the  same `--serve-params` json file for startup comparison between configurations would be really helpful (this can be a follow-up as well).

desertfire committed:
Add a script to benchmark startup time

Rename compilation benchmark to startup benchmark and measure total
startup time in addition to compilation time:

- Rename `vllm bench compilation` to `vllm bench startup`
- Measure both cold and warm startup times
- Report total startup time and compilation time separately
- Print percentiles for both startup time and compilation time

Signed-off-by: Bin Bao <binbao@meta.com>

desertfire committed:
Simplify startup.py

Signed-off-by: Bin Bao <binbao@meta.com>

ProExpertProg reviewed (approved):
Will try to review the follow-ups once I'm back! My definition of total startup time would be the time it takes for `vllm serve` to be ready to accept incoming requests.

zou3519 commented on vllm/benchmarks/startup.py:
```diff
@@ -0,0 +1,326 @@
+                hasattr(vllm_config, "compilation_config")
+                and vllm_config.compilation_config is not None
+            ):
+                compilation_time = vllm_config.compilation_config.compilation_time
```
related: I'm really not completely sure this time is correct. But it is the best thing we have. Also, how difficult would it be to have breakdowns by component?

(nothing needs to happen in this PR)

zou3519 replied:
See https://github.com/vllm-project/vllm/issues/27815

DarkLight1337 committed:
Merge branch 'main' into compile_time

desertfire commented:
> Getting error, we should add this script to CI:

What was the commandline you used to trigger the error? I don't run into any problem locally.

ProExpertProg commented:
> > Getting error, we should add this script to CI:
> 
> What was the commandline you used to trigger the error? I don't run into any problem locally.

I believe it was fixed in #31348.
